block_conv.go

```diff
@@ -0,0 +1,359 @@
+package slack
+
+import "encoding/json"
+
+// Marshalling/unmarshalling logic for Blocks
+
+// UnmarshalJSON implements the Unmarshaller interface for Blocks, so that any JSON
+// unmarshalling is delegated and proper type determination can be made before unmarshal
+func (b *Blocks) UnmarshalJSON(data []byte) error {
+	var raw []json.RawMessage
+	err := json.Unmarshal(data, &raw)
+	if err != nil {
+		return err
+	}
+
+	for _, r := range raw {
```

Comment:
this loses ordering of the blocks doesn't it?

the blocks type seems unncessary we want this function but have it return an array of the Block interface.
```golang
func foo(data []byte) (blocks []block, err error) {
        type union struct {
           Type string `json:"type"`
        }
	err := json.Unmarshal(data, &raw)
	if err != nil {
		return err
	}
        for _, r := range raw {
            u := union{}
            json.Unmarshal(r, &u)

            switch(u.Type) {
            case "actions":
                var tmp ActionBlock
                json.Unmarshal(r, &tmp)
                blocks = append(blocks, tmp)
            }
        }
        return blocks, nil
}
```

Reply:
@james-lawrence if we return an array of the block interface, how do you actually access the fields in that block? I want to be able to unmarshal the callback response, and then access the data in it for subsequent tasks, hence the `Blocks` type

Reply:
type switch. we're effectively talking about sum types here. which generally use type switches to determine the current type of an element.

I think its more important to maintain order of the messages than it is for each to be explicitly typed.

Reply:
@james-lawrence that's a fair point about ordering that I hadn't really considered. I suppose if you wanted to replace a message with the same block order + some small modifications, it would be pretty inconvenient to not know the original order of the blocks.

just to confirm, you are suggesting that library users implement the type switch themselves directly, yes?

Reply:
will update the PR after confirming this point and ask for re-review 👍

Reply:
Yes, I'm suggesting that library users implement the type switch themselves as then know exactly how they want to treat the blocks.

Reply:
All right I will adapt this and leave some examples on how to unmarshal, thanks for the response